fineract-progressive-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/transactionprocessor/impl/AdvancedPaymentScheduleTransactionProcessor.java

```diff
@@ -184,20 +184,48 @@ public Pair<ChangedTransactionDetail, ProgressiveLoanInterestScheduleModel> repr
         ProgressiveTransactionCtx ctx = new ProgressiveTransactionCtx(currency, installments, charges, overpaymentHolder,
                 changedTransactionDetail, scheduleModel);
 
-        for (final ChargeOrTransaction chargeOrTransaction : chargeOrTransactions) {
-            chargeOrTransaction.getLoanTransaction().ifPresent(loanTransaction -> processSingleTransaction(loanTransaction, ctx));
-            chargeOrTransaction.getLoanCharge()
+        final var loanTermVariations = Optional
+                .ofNullable(loan.getActiveLoanTermVariations()).map(loanTermVariationsSet -> loanTermVariationsSet.stream()
+                        .map(LoanTermVariations::toData).collect(Collectors.toCollection(ArrayList::new)))
+                .map(LoanTermVariationsDataWrapper::new).orElse(null);
+        final List<ChangeOperation> changeOperations = createSortedChangesList(loanTermVariations, loanTransactions, charges);
+        for (final ChangeOperation changeOperation : changeOperations) {
+            changeOperation.getInterestRateChange().ifPresent((interestRateChange) -> {
+                final LocalDate interestRateChangeEffectiveDate = interestRateChange.getTermVariationApplicableFrom().minusDays(1);
```

Comment:
Why `minusDays(1)`?

Reply:
because you will send it the next day (f.e. 15 February), but it counts from the 14 February

Reply:
in other words, if the interest change date is tomorrow then counts from today and doesn't affect the past

Reply:
Also would you mind to extract into a method? Similarly as `processSingleTransaction` or `processSingleCharge`?

Reply:
The installment reprocessing part has been extracted and I can move this logic to another coordinator method, 
but the work is still in progress on this part of the processor. So I can take care of the next phase.

Reply:
now already done this part also